# Import

In [12]:
import csv
import json
import pandas as pd
import requests
from skimpy import skim
import os

# Read in Data

In [19]:
os.getcwd()

'/Users/libbyseline/Documents/Winter25/capp122_clean/30122-project-gitmoney/data_smells'

In [20]:
raw = pd.read_csv("../data/jakejsmith_ChicagoMenuMoney_2012_23.csv")

FileNotFoundError: [Errno 2] No such file or directory: '../data/jakejsmith_ChicagoMenuMoney_2012_23.csv'

# Understanding Data
## Menu Money



In [103]:
skim(raw)


╭──────────────────────────────────────────────── skimpy summary ─────────────────────────────────────────────────╮
│          Data Summary                Data Types                                                                 │
│ ┏━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓ ┏━━━━━━━━━━━━━┳━━━━━━━┓                                                          │
│ ┃ Dataframe         ┃ Values ┃ ┃ Column Type ┃ Count ┃                                                          │
│ ┡━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩ ┡━━━━━━━━━━━━━╇━━━━━━━┩                                                          │
│ │ Number of rows    │ 28108  │ │ string      │ 4     │                                                          │
│ │ Number of columns │ 6      │ │ int64       │ 2     │                                                          │
│ └───────────────────┴────────┘ └─────────────┴───────┘                                                          │
│                                                     number                                                      │
│ ┏━━━━━━━━━━━━┳━━━━━━┳━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━┓  │
│ ┃ column     ┃ NA   ┃ NA %    ┃ mean     ┃ sd       ┃ p0      ┃ p25    ┃ p50    ┃ p75    ┃ p100   ┃ hist     ┃  │
│ ┡━━━━━━━━━━━━╇━━━━━━╇━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━┩  │
│ │ year       │    0 │       0 │     2018 │    3.415 │    2012 │   2015 │   2018 │   2021 │   2023 │  ▇▇████  │  │
│ │ ward       │    0 │       0 │    25.82 │    14.21 │       1 │     13 │     26 │     38 │     50 │  ▆█▇▇▆█  │  │
│ └────────────┴──────┴─────────┴──────────┴──────────┴─────────┴────────┴────────┴────────┴────────┴──────────┘  │
│                                                     string                                                      │
│ ┏━━━━━━━━━━━┳━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┓  │
│ ┃           ┃     ┃          ┃          ┃           ┃          ┃           ┃ chars    ┃ words per ┃ total    ┃  │
│ ┃ column    ┃ NA  ┃ NA %     ┃ shortest ┃ longest   ┃ min      ┃ max       ┃ per row  ┃ row       ┃ words    ┃  │
│ ┡━━━━━━━━━━━╇━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━┩  │
│ │ cost      │   0 │        0 │ $0.00    │ $1,152,00 │ $0.00    │ -$463.80  │     9.11 │         1 │    28108 │  │
│ │           │     │          │          │ 0.00      │          │           │          │           │          │  │
│ │ category  │   0 │        0 │ Lighting │ Plants,   │ Beautifi │ Streets & │     22.6 │       2.8 │    79101 │  │
│ │           │     │          │          │ Gardens,  │ cation   │ Transport │          │           │          │  │
│ │           │     │          │          │ &         │          │ ation     │          │           │          │  │
│ │           │     │          │          │ Sustainab │          │           │          │           │          │  │
│ │           │     │          │          │ ility     │          │           │          │           │          │  │
│ │ program   │   3 │ 0.010673 │ cm       │ 2 POD     │ 1 -      │ cmada     │     24.1 │       3.5 │    98536 │  │
│ │           │     │ 11797353 │          │ CAMERA -  │ LPR/Exis │           │          │           │          │  │
│ │           │     │     0667 │          │ 71st &    │ ting     │           │          │           │          │  │
│ │           │     │          │          │ Martin    │ Camera - │           │          │           │          │  │
│ │           │     │          │          │ Luther    │ NE Devon │           │          │           │          │  │
│ │           │     │          │          │ King Dr E │ and      │           │          │           │          │  │
│ │           │     │          │          │ 71ST ST & │ Ridge    │           │          │           │          │  │
│ │           │     │          │          │ S DR      │ 

**Do we have null values?**

From the above, we can tell most of our variables are filled in — with the exception of 109 descriptions and 3 programs. Both very small portions of the data. Wards, years, and categories are also complete, which will make our lives easier!

**What are our identifying columns?**

There really aren't any identifying columns ... let's see if we have duplicates. 

In [104]:
x = pd.DataFrame(raw.value_counts())
dupe_check = x.query("count > 1")

dupe_check

count
year ward cost        category                 program                                            description                                              
2017 47   $350.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... N CLARK ST & W AINSLIE ST                               3
          $600.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... W ADDISON ST & N OAKLEY AVE                             3
                                                                                                  W MONTROSE AVE & N HAMILTON AVE                         3
     1    $600.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... N CALIFORNIA AVE & W SHAKESPEARE AVE                    2
     47   $600.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... N LINCOLN AVE & W LELAND AVE                            2
                                                                                                  W FOSTER AVE & N WAYNE AVE                              2
2013 31   $950.00     Streets & Transportation In-Road 'State Law Stop For Pedestrians' Signs ... N CENTRAL AVE & W GEORGE ST                             2
2015 8    $156,000.00 Lighting                 Street Light Residential Staggered Piggy Back Menu LED Residential w/piggybacks. Current W.O # : 1...      2
2013 42   $550.00     Streets & Transportation In-Road 'State Law Stop For Pedestrians' Signs ... N WELLS ST & W WALTON ST                                2
     7    $550.00     Streets & Transportation In-Road 'State Law Stop For Pedestrians' Signs ... E 93RD ST & S OGLESBY AVE                               2
2018 35   $350.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... N MILWAUKEE AVE & N SAWYER AVE                          2
2017 1    $350.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... W DIVISION ST & N PAULINA ST                            2
2018 36   $600.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... NRRAGANSETT AVE & W BARRY AVE                           2
2017 1    $350.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... W MEDILL AVE & N MILWAUKEE AVE                          2
          $600.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... N CALIFORNIA AVE & W ALTGELD ST                         2
     48   $0.00       Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... N CLARK ST & W OLIVE AVE                                2
     1    $600.00     Streets & Transportation In-Road "State Law Stop For Pedestrians" Signs ... N DAMEN AVE & W HURON ST                                2

In [105]:
dupe_check.shape

(17, 1)

**Answer**: We have 17 duplicate rows. 

**Are there immediate outliers?**

I guess to do that, we should look at the cost. But as long as it's less than 1.5 mil -- we should be OK

To do that, first we should tranform the numbers from a string to a number. We may have to remove the dollar sign. 

In [106]:
raw['cost'] = raw['cost'].str.replace('$', '')
raw['cost'] = raw['cost'].str.replace(',', '')
raw

year  ward       cost                  category  \
0      2012     1    6354.00  Streets & Transportation   
1      2012     1    2800.00  Streets & Transportation   
2      2012     1    3150.00  Streets & Transportation   
3      2012     1   19882.00  Streets & Transportation   
4      2012     1    1000.00  Streets & Transportation   
...     ...   ...        ...                       ...   
28103  2023    50   38695.82  Streets & Transportation   
28104  2023    50  134804.82  Streets & Transportation   
28105  2023    50   54020.66  Streets & Transportation   
28106  2023    50       0.00  Streets & Transportation   
28107  2023    50       0.00  Streets & Transportation   

                                    program  \
0                    Alley Resurfacing Menu   
1                     Alley Speed Hump Menu   
2                              Bollard Menu   
3                        Curb & Gutter Menu   
4                         Pavement Markings   
...                                     ...   
28103             Street Resurfacing (2023)   
28104             Street Resurfacing (2023)   
28105             Street Resurfacing (2023)   
28106  Street Speed Hump Replacement (2023)   
28107  Street Speed Hump Replacement (2023)   

                                             description  
0      N WESTERN AVE & N CLAREMONT AVE & W NORTH AVE ...  
1      N WOOD ST & N HONORE ST & W AUGUSTA BLVD & W I...  
2      ON W BELMONT AVE FROM 2441 W TO N CAMPBELL AV ...  
3      ON N MAPLEWOOD AVE FROM W BELDEN AV (2300 N) T...  
4                        W CHARLESTON ST & N WESTERN AVE  
...                                                  ...  
28103  ON N WINCHESTER AVE FROM 6445 N TO W LOYOLA AV...  
28104  ON W BIRCHWOOD AVE FROM N ROCKWELL ST (2600 W)...  
28105  ON W JARVIS AVE FROM N FRANCISCO AVE (2900 W) ...  
28106  ON N MOZART ST FROM W DEVON AVE (6400 N) TO W ...  
28107  ON W JARVIS AVE FROM N FRANCISCO AVE (2900 W) ...  

[28108 rows x 6 columns]

In [109]:
raw['cost'] = pd.to_numeric(raw['cost'])

skim(raw)

╭──────────────────────────────────────────────── skimpy summary ─────────────────────────────────────────────────╮
│          Data Summary                Data Types                                                                 │
│ ┏━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓ ┏━━━━━━━━━━━━━┳━━━━━━━┓                                                          │
│ ┃ Dataframe         ┃ Values ┃ ┃ Column Type ┃ Count ┃                                                          │
│ ┡━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩ ┡━━━━━━━━━━━━━╇━━━━━━━┩                                                          │
│ │ Number of rows    │ 28108  │ │ string      │ 3     │                                                          │
│ │ Number of columns │ 6      │ │ int64       │ 2     │                                                          │
│ └───────────────────┴────────┘ │ float64     │ 1     │                                                          │
│                                └─────────────┴───────┘                                                          │
│                                                     number                                                      │
│ ┏━━━━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━┓  │
│ ┃ column    ┃ NA   ┃ NA %   ┃ mean    ┃ sd      ┃ p0       ┃ p25    ┃ p50     ┃ p75     ┃ p100      ┃ hist   ┃  │
│ ┡━━━━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━┩  │
│ │ year      │    0 │      0 │    2018 │   3.415 │     2012 │   2015 │    2018 │    2021 │      2023 │ ▇▇████ │  │
│ │ ward      │    0 │      0 │   25.82 │   14.21 │        1 │     13 │      26 │      38 │        50 │ ▆█▇▇▆█ │  │
│ │ cost      │    0 │      0 │   28600 │   44700 │   -14250 │   2459 │   11760 │   40590 │   1152000 │   █    │  │
│ └───────────┴──────┴────────┴─────────┴─────────┴──────────┴────────┴─────────┴─────────┴───────────┴────────┘  │
│                                                     string                                                      │
│ ┏━━━━━━━━━━━┳━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┓  │
│ ┃           ┃     ┃          ┃          ┃           ┃          ┃           ┃ chars    ┃ words per ┃ total    ┃  │
│ ┃ column    ┃ NA  ┃ NA %     ┃ shortest ┃ longest   ┃ min      ┃ max       ┃ per row  ┃ row       ┃ words    ┃  │
│ ┡━━━━━━━━━━━╇━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━┩  │
│ │ category  │   0 │        0 │ Lighting │ Plants,   │ Beautifi │ Streets & │     22.6 │       2.8 │    79101 │  │
│ │           │     │          │          │ Gardens,  │ cation   │ Transport │          │           │          │  │
│ │           │     │          │          │ &         │          │ ation     │          │           │          │  │
│ │           │     │          │          │ Sustainab │          │           │          │           │          │  │
│ │           │     │          │          │ ility     │          │           │          │           │          │  │
│ │ program   │   3 │ 0.010673 │ cm       │ 2 POD     │ 1 -      │ cmada     │     24.1 │       3.5 │    98536 │  │
│ │           │     │ 11797353 │          │ CAMERA -  │ LPR/Exis │           │          │           │          │  │
│ │           │     │     0667 │          │ 71st &    │ ting     │           │          │           │          │  │
│ │           │     │          │          │ Martin    │ Camera - │           │          │           │          │  │
│ │           │     │          │          │ Luther    │ NE Devon │           │          │           │          │  │
│ │           │     │          │          │ King Dr E │ and      │           │          │           │          │  │
│ │           │     │          │          │ 71ST ST & │ Ridge    │           │          │           │          │  │
│ │           │     │          │          │ S DR      │ 

Ope that looks like a negative cost ... that's not good

In [112]:
raw.query("cost < 0")

year  ward      cost                  category  \
17874  2019    38 -14253.07  Streets & Transportation   
20424  2020    48   -463.80  Streets & Transportation   

                                                 program  \
17874                          Street Resurfacing (2019)   
20424  Street Resurfacing (2020) with cost adjustment...   

                                             description  
17874  ON N ORIOLE AVE FROM W CORNELIA AVE (3499 N) T...  
20424  ON W GLENLAKE AVE FROM N KENMORE AVE (1040 W) ...

Okay, two negative costs. Both with street resurfacing.

Checked Street Resurfacing Menu (2019) ON N ORIOLE AVE FROM W CORNELIA AVE (3499 N)
- that is indeed a negative number
Checked Street Resurfacing (2020) ON W GLENLAKE AVE FROM N KENMORE AVE (1040 W)
- that is indeed a negative number

**What's the universe of the data? Why and how is it collected?**

**Do we know what every column means?**